File: Stack/20有效的括号.cpp
#include <stack>
#include <string>
using namespace std;
class Solution {
public:
    bool isValid(string s) {
        stack<char>astack;
        for(char now:s){
            if(now=='{'||now=='['||now=='('){
                astack.push(now);
            }
            else{
                if(astack.empty())return false;
                if(now=='}'){
                    if(astack.top()!='{')return false;
                    astack.pop();
                }
                else if(now==']'){
                    if(astack.top()!='[')return false;
                    astack.pop();
                }
                else if(now==')'){
                    if(astack.top()!='(')return false;
                    astack.pop();
                }
            }
        }
        if(!astack.empty())return false;
        return true;
    }
};

File: Stack/20有效的括号.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for now in s:
            if now == '[' or now == '{' or now == '(':
                stack.append(now)
                continue
            if now == ']':
                if len(stack) == 0 or stack[-1] !='[':
                    return False
            if now == '}':
                if len(stack) == 0 or stack[-1] != '{':
                    return False
            if now == ')':
                if len(stack) == 0 or stack[-1] != '(':
                    return False
            stack.pop()
        if len(stack) != 0:
            return False
        return True